token.go

```diff
@@ -0,0 +1,232 @@
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+	signed := (*cms.ParsedSignedData)(t)
+	var verifiedCerts []*x509.Certificate
+	var lastErr error
+	for _, signerInfo := range t.SignerInfos {
+		signingCertificate, err := t.GetSigningCertificate(ctx, &signerInfo)
+		if err != nil {
+			lastErr = SignedTokenVerificationError{Detail: err}
+			continue
+		}
+		if _, err := signed.VerifySigner(ctx, &signerInfo, signingCertificate, opts); err != nil {
+			lastErr = SignedTokenVerificationError{Detail: err}
+			continue
+		}
+		// RFC 3161 2.3: The corresponding certificate MUST contain only one
+		// instance of the extended key usage field extension.
+		if len(signingCertificate.ExtKeyUsage) == 1 &&
+			signingCertificate.ExtKeyUsage[0] == x509.ExtKeyUsageTimeStamping &&
+			len(signingCertificate.UnknownExtKeyUsage) == 0 {
+			verifiedCerts = append(verifiedCerts, signingCertificate)
+		} else {
+			lastErr = SignedTokenVerificationError{Msg: "signing certificate MUST only have ExtKeyUsageTimeStamping as extended key usage"}
+		}
+	}
+	if len(verifiedCerts) == 0 {
+		return nil, lastErr
+	}
+	return verifiedCerts, nil
+}
+
+// Info returns the timestamping information.
+func (t *SignedToken) Info() (*TSTInfo, error) {
+	var info TSTInfo
+	if _, err := asn1.Unmarshal(t.Content, &info); err != nil {
+		return nil, err
+	}
+	// RFC 3161 2.4.2 TSTInfo
+	if info.Version != 1 {
+		return nil, fmt.Errorf("timestamp token info version must be 1, but got %d", info.Version)
+	}
+	return &info, nil
+}
+
+// GetSigningCertificate gets the signing certificate identified by SignedToken
+// SignerInfo's SigningCertificate attribute. If the IssuerSerial
+// field of signing certificate is missing, use signerInfo's sid instead.
+// The identified signing certificate MUST match the hash in SigningCertificate.
+//
+// References: RFC 3161 2.4.1 & 2.4.2; RFC 5035 4
+func (t *SignedToken) GetSigningCertificate(ctx context.Context, signerInfo *cms.SignerInfo) (*x509.Certificate, error) {
+	var signingCertificateV2 SigningCertificateV2
+	if err := signerInfo.SignedAttributes.TryGet(oid.SigningCertificateV2, &signingCertificateV2); err != nil {
+		return nil, fmt.Errorf("failed to get SigningCertificateV2 from signed attributes: %w", err)
+	}
+	// get candidate signing certificate
+	var candidateSigningCert *x509.Certificate
+	signed := (*cms.ParsedSignedData)(t)
+	if signingCertificateV2.Certificates[0].IssuerSerial.Issuer.FullBytes != nil {
+		// use IssuerSerial from signingCertificateV2
+		var rawValue asn1.RawValue
+		_, err := asn1.Unmarshal(signingCertificateV2.Certificates[0].IssuerSerial.Issuer.Bytes, &rawValue)
+		if err != nil {
+			return nil, err
+		}
+		// RFC 3280 4.2.1.7
+		if rawValue.Tag != 4 {
+			return nil, fmt.Errorf("signing certificate IssuerSerial requires CHOICE tag 4, but got %d", rawValue.Tag)
+		}
+		var issuer asn1.RawValue
+		_, err = asn1.Unmarshal(rawValue.Bytes, &issuer)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
We can simplify this part:
response.go
```go
type issuerAndSerial struct {
	IssuerName   generalNames
	SerialNumber *big.Int
}

type generalNames struct {
	Name asn1.RawValue `asn1:"optional,tag:4"`
}

// ESSCertIDv2 uniquely identifies a certificate.
//
// Reference: RFC 5035 4 ESSCertIDv2
//
//	ESSCertIDv2 ::=  SEQUENCE {
//	 hashAlgorithm           AlgorithmIdentifier
//	 	DEFAULT {algorithm id-sha256},
//	 certHash                 Hash,
//	 issuerSerial             IssuerSerial OPTIONAL }
type ESSCertIDv2 struct {
	// HashAlgorithm is the hashing algorithm used to hash certificate.
	// When it is not present, the default value is SHA256 (id-sha256).
	// SHA1 is unsupported.
	HashAlgorithm pkix.AlgorithmIdentifier `asn1:"optional"`

	// CertHash is the certificate hash using algorithm identified
	// by HashAlgorithm. It is is computed over the entire DER-encoded
	// certificate (including the signature)
	CertHash []byte

	// IssuerSerial holds the issuer and serialNumber of the certificate
	// When it is not present, the SignerIdentifier field in the SignerInfo
	// will be used.
	IssuerSerial issuerAndSerial `asn1:"optional"`
}
```

```suggestion
		var issuer asn1.RawValue
		_, err := asn1.Unmarshal(signingCertificateV2.Certificates[0].IssuerSerial.IssuerName.Name.Bytes, &issuer)
		if err != nil {
			return nil, err
		}
```

Then, if `IssuerSerial.IssuerName.Name`'s tag is not 4, `IssuerSerial.IssuerName.Name` will be empy field.

Reply:
That's a good idea! (although I made some changes to make it work)